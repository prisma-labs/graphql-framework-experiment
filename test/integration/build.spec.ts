// TODO boot and query to take integration test confidence even further

import { createWorkspace } from '../__helpers'

const ws = createWorkspace({
  name: 'build',
})

it('can build with just a schema module', () => {
  ws.fs.write(
    'schema.ts',
    `
      objectType({
        name: 'A',
        definition(t) {
          t.string('a')
        }
      })
    `
  )

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})

it('can build with just a schema folder of modules', () => {
  ws.fs.write(
    'schema/a.ts',
    `
      objectType({
        name: 'A',
        definition(t) {
          t.string('a')
        }
      })
    `
  )

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})

it('can build with schema + app modules', () => {
  ws.fs.write(
    'schema.ts',
    `
      objectType({
        name: 'A',
        definition(t) {
          t.string('a')
        }
      })
    `
  )

  ws.fs.write('app.ts', `app.server.start()`)

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})

it('can nest modules', () => {
  ws.fs.write(
    'graphql/schema.ts',
    `
      objectType({
        name: 'A',
        definition(t) {
          t.string('a')
        }
      })
    `
  )

  ws.fs.write('graphql/app.ts', `app.server.start()`)

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})

it('can build a plugin', () => {
  ws.fs.write(
    'myplugin.ts',
    `	
      import { Plugin } from 'pumpkins'	

      export type Context = {	
        a: 1	
      }	
      	
      export default {	
        name: 'myplugin',	
        context: {	
          typeSourcePath: __filename,	
          create() {	
            return {	
              a: 1,	
            }	
          },	
        },	
      } as Plugin<Context>	
    `
  )

  ws.fs.write(
    'schema.ts',
    `queryType({	
        definition(t) {	
          t.int('a', (_root, _args, ctx) => ctx.a)	
          t.field('foo', {	
            type: 'Foo',	
            resolve() {	
              return {	
                bar: 'qux'	
              }	
            }	
          })	
        }	
      })	

      objectType({	
        name: 'Foo',	
        definition(t) {	
          t.string('bar')	
        }	
      })	
    `
  )

  ws.fs.write(
    'app.ts',
    `	
      import myplugin from './myplugin'	
      app.use(myplugin).server.start()	
    `
  )

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})

it.skip('can build a prisma framework project', () => {
  ws.fs.write(
    'schema.prisma',
    `	
      datasource db {	
        provider = "sqlite"	
        url      = "file:dev.db"	
      }	
      generator photon {	
        provider = "photonjs"	
      }	
      model User {	
        id   Int    @id	
        name String	
      }	
    `
  )

  ws.fs.write(
    'schema.ts',
    `
      objectType({
        name: 'User',
        definition(t) {
          t.model.id()
          t.model.name()
        }
      })
    `
  )

  const result = ws.run('yarn -s pumpkins build')
  expect(result).toMatchSnapshot()
  expect(ws.fs.inspectTree('build')).toMatchSnapshot()
})
