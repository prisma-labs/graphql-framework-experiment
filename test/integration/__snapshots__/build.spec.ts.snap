// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can build a plugin 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "ðŸŽƒ  Generating Nexus artifacts ...
ðŸŽƒ  Compiling ...
ðŸŽƒ  Pumpkins server successfully compiled!
",
}
`;

exports[`can build a plugin 2`] = `
Object {
  "children": Array [
    Object {
      "name": "app.js",
      "size": 319,
      "type": "file",
    },
    Object {
      "name": "myplugin.js",
      "size": 268,
      "type": "file",
    },
    Object {
      "name": "schema.js",
      "size": 366,
      "type": "file",
    },
    Object {
      "name": "start.js",
      "size": 625,
      "type": "file",
    },
  ],
  "name": "dist",
  "size": 1578,
  "type": "dir",
}
`;

exports[`can build with just a schema folder of modules 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "ðŸŽƒ  Generating Nexus artifacts ...
ðŸŽƒ  Compiling ...
ðŸŽƒ  Pumpkins server successfully compiled!
",
}
`;

exports[`can build with just a schema folder of modules 2`] = `
Object {
  "children": Array [
    Object {
      "name": "a.js",
      "size": 95,
      "type": "file",
    },
    Object {
      "name": "start.js",
      "size": 472,
      "type": "file",
    },
  ],
  "name": "build",
  "size": 567,
  "type": "dir",
}
`;

exports[`can build with just a schema module 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "ðŸŽƒ  Generating Nexus artifacts ...
ðŸŽƒ  Compiling ...
ðŸŽƒ  Pumpkins server successfully compiled!
",
}
`;

exports[`can build with just a schema module 2`] = `
Object {
  "children": Array [
    Object {
      "name": "schema.js",
      "size": 95,
      "type": "file",
    },
    Object {
      "name": "start.js",
      "size": 470,
      "type": "file",
    },
  ],
  "name": "build",
  "size": 565,
  "type": "dir",
}
`;

exports[`can build with schema + app modules 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "ðŸŽƒ  Generating Nexus artifacts ...
ðŸŽƒ  Compiling ...
ðŸŽƒ  Pumpkins server successfully compiled!
",
}
`;

exports[`can build with schema + app modules 2`] = `
Object {
  "children": Array [
    Object {
      "name": "app.js",
      "size": 34,
      "type": "file",
    },
    Object {
      "name": "schema.js",
      "size": 95,
      "type": "file",
    },
    Object {
      "name": "start.js",
      "size": 625,
      "type": "file",
    },
  ],
  "name": "dist",
  "size": 754,
  "type": "dir",
}
`;
